### ISSUE TITLE
Exceptions that are not caused by the VM should terminate program execution

### ISSUE THREAD

**Issue by jwasinger** [ethereumjs/ethereumjs-vm#191]
Right now, all exceptions are caught and interpreted as errors within the EVM environment (i.e. out of gas).  This causes a myriad of problems.

**Comment by kumavis** [comment]
yeah this is really annoying
any non-vm error should be considered a catastrophic event and fail immediately

err / vm-err is conflated here https://github.com/ethereumjs/ethereumjs-vm/blob/master/lib/runCode.js#L217
err is lost here https://github.com/ethereumjs/ethereumjs-vm/blob/master/lib/runCall.js#L154-L185

**Comment by kumavis** [comment]
old duplicate https://github.com/ethereumjs/ethereumjs-vm/issues/117

**Comment by holgerd77** [comment]
Jared once made the following suggestion when we were chatting on Gitter:


> I'm thinking we add special field to mark EVM exceptions.
> then in that try/catch
> it will look something like this:

```
try {

} catch (e) {
    if typeof(e) == 'object ' && 'EVMException' in Object.keys(e) {
        OOG Handling...
    }  else {
        throw("real exception, which ends program excecution")
   }
}
```

@jwasinger: is this still up-to-date or do you have updated ideas on this?

**Comment by axic** [comment]
#222 is one step towards that (introduces `internal_error`)

**Comment by holgerd77** [comment]
@axic I think it would be better to keep the vm errors logically separate from programmatic errors. Any reasons not to settle this on one level higher and introduce a new ``runState.internalError`` variable (with some adoptions to post-handling)? 

Then one gets also more flexible to store the whole error with the stack trace.

**Comment by axic** [comment]
@holgerd77 now that #219 is merged it should be easy to change whatever needs to be changed to return `INTERNAL_ERROR` instead. #222 will introduce internal errors for stack mismatch (if the opcode returns an inconsistent count).

**Comment by axic** [comment]
I think this should be possible now since we have `INTERNAL_ERROR` signaling a caught exception.

**Comment by holgerd77** [comment]
@axic Yes, can do.

**Comment by axic** [comment]
Fixed by #307.
